Issue 101: Wrap code inside _process_data()

adusca opened the pull request:
I moved computing the dictionaries to the function _process_data(). Also, instead of using all_builders_information we are now just calling fetch_allthethings_data(). To avoid making things too slow I added some in-memory-caching to allthethings.py. The final result is actually faster then before this patch (py.test went from 16s to 7s on my machine)

armenzg commented on mozci/platforms.py:
```diff
@@ -51,44 +45,47 @@ def is_downstream(buildername):
 # every test job. In this case, "Android 4.0 armv7 API 11+ larch opt
 # test cppunit" : larch-android-api-11-opt-unittest
 buildername_to_trigger = {}
```
Could you please upper case this variable? This is to meet higher pep8 score.

armenzg commented on mozci/platforms.py:
```diff
@@ -51,44 +45,47 @@ def is_downstream(buildername):
-    for buildername in values['downstream']:
-        assert buildername not in buildername_to_trigger
-        buildername_to_trigger[buildername.lower()] = values['triggered_by'][0]
+build_jobs = {}
```
Could you please upper case this variable? This is to meet higher pep8 score.

armenzg commented on mozci/platforms.py:
```diff
@@ -51,44 +45,47 @@ def is_downstream(buildername):
+    '''Filling the dictionaries used by determine_upstream_builder. '''
+    # We check if we already computed before
+    if buildername_to_trigger:
+        return
```
Could you please add a debug message?

armenzg commented on mozci/platforms.py:
```diff
@@ -51,44 +45,47 @@ def is_downstream(buildername):
+
+    # We'll look at every builder and if it's a build job we will add it
+    # to shortname_to_name
+    for buildername, builderinfo in fetch_allthethings_data()['builders'].iteritems():
```
Could you please add a debug message that we're running this?

armenzg commented on mozci/sources/allthethings.py:
```diff
@@ -58,6 +58,8 @@
 ALLTHETHINGS = \
     "https://secure.pub.build.mozilla.org/builddata/reports/allthethings.json"
 
+_data = {}
```
Could you please upper case this variable? This is to meet higher pep8 score.

armenzg commented on mozci/sources/allthethings.py:
```diff
@@ -93,21 +95,25 @@ def _verify_file_integrity():
 
+    global _data
     if no_caching:
-        data = _fetch()
```
I realized that we can improve the readability of this block by doing this:
> if no_caching or not os.path.exists(FILENAME)
since both cases do _data = _fetch()

armenzg commented on mozci/sources/allthethings.py:
```diff
@@ -93,21 +95,25 @@ def _verify_file_integrity():
                 fd = open(FILENAME)
-                data = json.load(fd)
+                _data = json.load(fd)
         else:
```
If we change the if/condition we can remove this "else" clause.

armenzg commented:
Alice, these are excellent news! I'm happy to hear that it makes such a big difference.
I should have caught this in our very first initial reviews. My apologies.

Feel free to merge once you address my comment. Obrigado!

adusca commented:
I addressed your comments, but I ended up changing more things in the fetch logic. I want a second opinion on these changes, so I'm not merging it immediately (but no hurry!).

armenzg commented on mozci/sources/allthethings.py:
```diff
@@ -84,6 +87,9 @@ def _fetch():
     def _verify_file_integrity():
+        if not os.path.exists(FILENAME):
+            return False
+
```
Don't do this. See in a comment below why.

armenzg commented on mozci/sources/allthethings.py:
```diff
@@ -93,21 +99,20 @@ def _verify_file_integrity():
-                data = json.load(fd)
+        DATA = _fetch()
+    # If we do not have an in-memory cache, try to use the file cache.
+    elif not DATA:
```
A more pythonic way of writing this is "elif DATA is not None".

Could you please fix that?

>>> a = {}
>>> print not a
True
>>> print a is not None
True
>>> print a == {}
True

armenzg commented on mozci/sources/allthethings.py:
```diff
@@ -93,21 +99,20 @@ def _verify_file_integrity():
+            DATA = json.load(fd)
         else:
-            data = _fetch()
+            DATA = _fetch()
```
We need to delete the file if the verification is incorrect.
We're loosing that with this new approach.

adusca commented:
Done!
